lxd/auth/authorization_objects.go

```diff
@@ -0,0 +1,208 @@
+	return fmt.Sprintf("%s:%s/%s/%s/%s", ObjectTypeStoragePoolVolume, url.PathEscape(projectName), url.PathEscape(poolName), url.PathEscape(volumeType), url.PathEscape(volumeName))
+}
+
+// ObjectByType returns an object of the given type. The passed in arguments must be in the correct
+// order (as found in the URL for the resource). This function will error if an invalid object type is
+// given, or if the correct number of arguments is not passed in. Project objects use the "projectName" argument.
+func ObjectByType(objectType ObjectType, projectName string, args ...string) (string, error) {
+	switch objectType {
+	case ObjectTypeServer:
+		return ObjectServer(), nil
+	case ObjectTypeCertificate:
+		if len(args) != 1 {
+			return "", fmt.Errorf("Missing certificate fingerprint argument")
+		}
+
+		return ObjectCertificate(args[0]), nil
+	case ObjectTypeStoragePool:
+		if len(args) != 1 {
+			return "", fmt.Errorf("Missing storage pool name argument")
+		}
+
+		return ObjectStoragePool(args[0]), nil
+	case ObjectTypeProject:
+		return ObjectProject(projectName), nil
+	case ObjectTypeImage:
+		if len(args) != 1 {
+			return "", fmt.Errorf("Missing image fingerprint argument")
+		}
+
+		return ObjectImage(projectName, args[0]), nil
+	case ObjectTypeImageAlias:
+		if len(args) != 1 {
+			return "", fmt.Errorf("Missing image alias name argument")
+		}
+
+		return ObjectImageAlias(projectName, args[0]), nil
+	case ObjectTypeInstance:
+		if len(args) != 1 {
+			return "", fmt.Errorf("Missing instance name argument")
+		}
+
+		return ObjectInstance(projectName, args[0]), nil
+	case ObjectTypeNetwork:
+		if len(args) != 1 {
+			return "", fmt.Errorf("Missing network name argument")
+		}
+
+		return ObjectNetwork(projectName, args[0]), nil
+	case ObjectTypeNetworkACL:
+		if len(args) != 1 {
+			return "", fmt.Errorf("Missing network ACL name argument")
+		}
+
+		return ObjectNetworkACL(projectName, args[0]), nil
+	case ObjectTypeNetworkZone:
+		if len(args) != 1 {
+			return "", fmt.Errorf("Missing network zone name argument")
+		}
+
+		return ObjectNetworkZone(projectName, args[0]), nil
+	case ObjectTypeProfile:
+		if len(args) != 1 {
+			return "", fmt.Errorf("Missing profile name argument")
+		}
+
+		return ObjectProfile(projectName, args[0]), nil
+	case ObjectTypeStorageBucket:
+		if len(args) != 2 {
+			return "", fmt.Errorf("Missing required arguments for a storage bucket object")
+		}
+
+		return ObjectStorageBucket(projectName, args[0], args[1]), nil
+	case ObjectTypeStoragePoolVolume:
+		if len(args) != 3 {
+			return "", fmt.Errorf("Missing required arguments for a storage pool volume object")
+		}
+
+		return ObjectStoragePoolVolume(projectName, args[0], args[1], args[2]), nil
+	}
+
+	return "", fmt.Errorf("Unknown object type %q", objectType)
+}
+
+// ObjectFromRequest returns an object created from the request by evaluating the given mux vars.
```

Comment:
The word "object" seems a bit misleading as this function returns a string. I realise Go doesn't have "objects" as such but often people refer to instantiated structs as objects. What is the object that this function returns? Can we clarify it by changing the type to a dedicated string type perhaps?

Reply:
I'll try to change things a bit to make things a bit more obvious. I'll put some examples in here first to get consensus though. If only because rebasing this thing is starting to get tricky :laughing:

Reply:
@tomponline I've added a WIP commit with some changes to this file to try and make things clearer. Can you please take a look at this and tell me what you think? If it looks good I'll make the change and rebase through both PRs. https://github.com/markylaing/lxd/blob/de51714968a469599a3ea272cf9433b2ccc57893/lxd/auth/authorization_objects.go

Reply:
You can also view the WIP commit to see how this change affects other parts of the codebase https://github.com/canonical/lxd/pull/12252/commits/de51714968a469599a3ea272cf9433b2ccc57893

Reply:
Yeah this looks good. I think it would be worth adding a comment to the `type Object string` definition explaining what it is though.

Reply:
Yep I will do. The WIP commit is failing CI at the minute so on Monday morning I'll fix those errors, fix up comments etc. and have a look at the rest of your comments 🙂

Reply:
Ta, I think the rest of the commits looked somewhat mechanical so ill wait until those comments have been addressed before reviewing rest of it.